File: backend/app/schemas/analytics.py
from datetime import date
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class MonthlyCategorySpendingRead(BaseModel):
    month: date
    category_name: str
    parent_category_name: str
    transaction_count: int
    total_spending: Decimal


class CreditUtilizationRead(BaseModel):
    account_id: UUID
    account_name: str
    bank_name: Optional[str] = None
    card_number_masked: str
    credit_limit: Decimal
    current_balance: Decimal
    available_limit: Decimal
    utilization_percentage: Decimal
    risk_level: str


class UpcomingObligationRead(BaseModel):
    obligation_type: str
    reference_id: UUID
    account_name: str
    card_number_masked: str
    due_date: date
    days_remaining: int
    total_amount_due: Decimal
    minimum_amount_due: Decimal
    payment_status: str


class NetWorthOverviewRead(BaseModel):
    total_liquid_assets: Decimal
    total_bank_assets: Decimal
    total_cash_assets: Decimal
    total_ewallet_assets: Decimal
    total_savings_assets: Decimal
    total_credit_debt: Decimal
    total_credit_limit: Decimal
    total_available_credit: Decimal
    net_worth: Decimal
    active_asset_accounts_count: int
    active_credit_cards_count: int


class MonthlyCashFlowRead(BaseModel):
    month: date
    total_income: Decimal
    total_expense: Decimal
    net_savings: Decimal
    savings_rate_percent: Decimal
    total_transactions_count: int


class DashboardOverviewRead(BaseModel):
    total_liquid_assets: Decimal = Decimal("0.00")
    total_bank_assets: Decimal = Decimal("0.00")
    total_cash_assets: Decimal = Decimal("0.00")
    total_ewallet_assets: Decimal = Decimal("0.00")
    total_credit_limit: Decimal = Decimal("0.00")
    total_live_balance: Decimal = Decimal("0.00")
    total_available_limit: Decimal = Decimal("0.00")
    net_worth: Decimal = Decimal("0.00")
    overall_utilization_percentage: Decimal = Decimal("0.00")
    overall_risk_level: str = "OPTIMAL (<30%)"
    active_cards_count: int = 0
    active_asset_accounts_count: int = 0
    upcoming_obligations_count: int = 0
    total_upcoming_due_30d: Decimal = Decimal("0.00")
    monthly_spending_current_month: Decimal = Decimal("0.00")
    monthly_income_current_month: Decimal = Decimal("0.00")

File: frontend/src/types/analytics.ts
export interface MonthlyCategorySpending {
  month: string;
  category_name: string;
  parent_category_name: string;
  transaction_count: number;
  total_spending: number;
}

export interface CreditUtilization {
  account_id: string;
  account_name: string;
  bank_name?: string;
  card_number_masked: string;
  credit_limit: number;
  current_balance: number;
  available_limit: number;
  utilization_percentage: number;
  risk_level: string;
}

export interface UpcomingObligation {
  obligation_type: "STATEMENT" | "INSTALLMENT";
  reference_id: string;
  account_name: string;
  card_number_masked: string;
  due_date: string;
  days_remaining: number;
  total_amount_due: number;
  minimum_amount_due: number;
  payment_status: string;
}

export interface NetWorthOverview {
  total_liquid_assets: number;
  total_bank_assets: number;
  total_cash_assets: number;
  total_ewallet_assets: number;
  total_savings_assets: number;
  total_credit_debt: number;
  total_credit_limit: number;
  total_available_credit: number;
  net_worth: number;
  active_asset_accounts_count: number;
  active_credit_cards_count: number;
}

export interface MonthlyCashFlow {
  month: string;
  total_income: number;
  total_expense: number;
  net_savings: number;
  savings_rate_percent: number;
  total_transactions_count: number;
}

export interface DashboardOverview {
  total_liquid_assets: number;
  total_bank_assets: number;
  total_cash_assets: number;
  total_ewallet_assets: number;
  total_credit_limit: number;
  total_live_balance: number;
  total_available_limit: number;
  net_worth: number;
  overall_utilization_percentage: number;
  overall_risk_level: string;
  active_cards_count: number;
  active_asset_accounts_count: number;
  upcoming_obligations_count: number;
  total_upcoming_due_30d: number;
  monthly_spending_current_month: number;
  monthly_income_current_month: number;
}
